analysis/complex.lean

```diff
@@ -0,0 +1,281 @@
+noncomputable theory
+-- because reals are noncomputable
+
+local attribute [instance] classical.decidable_inhabited classical.prop_decidable
+-- because I don't know how to do inverses sensibly otherwise;
+-- e.g. I needed to know that if z was non-zero then either its real part
+-- was non-zero or its imaginary part was non-zero.
+
+structure complex : Type :=
+(re : ℝ) (im : ℝ)
+
+notation `ℂ` := complex
+
+-- definition goes outside namespace, then everything else in it?
+
+namespace complex
+
+-- handy checks for equality etc
+
+theorem eta (z : complex) : complex.mk z.re z.im = z :=
+  cases_on z (λ _ _, rfl)
+
+-- very useful
+
+theorem eq_of_re_eq_and_im_eq (z w : complex) : z.re=w.re ∧ z.im=w.im → z=w :=
+begin
+intro H,rw [←eta z,←eta w,H.left,H.right],
+end
+
+-- simp version
+
+theorem eq_iff_re_eq_and_im_eq (z w : complex) : z=w ↔ z.re=w.re ∧ z.im=w.im :=
+begin
+split,
+  intro H,rw [H],split;trivial,
+exact eq_of_re_eq_and_im_eq _ _,
+end
+
+lemma proj_re (r0 i0 : real) : (complex.mk r0 i0).re = r0 := rfl
+lemma proj_im (r0 i0 : real) : (complex.mk r0 i0).im = i0 := rfl
+
+local attribute [simp] eq_iff_re_eq_and_im_eq proj_re proj_im
+
+-- Am I right in
+-- thinking that the end user should not need to
+-- have to use this function?
+
+def of_real : ℝ → ℂ := λ x, { re := x, im := 0 }
+
+-- does one name these instances or not? I've named a random selection
+
+instance coe_real_complex : has_coe ℝ ℂ := ⟨of_real⟩
+instance : has_zero complex := ⟨of_real 0⟩
+instance : has_one complex := ⟨of_real 1⟩
+instance inhabited_complex : inhabited complex := ⟨0⟩
+
+-- dangerously short variable name so I protected it.
+-- It's never used in the library (other than in the projection
+-- commands) but I think end users will use it.
+
+protected def i : complex := {re := 0, im := 1}
+
+def conjugate (z : complex) : complex := {re := z.re, im := -(z.im)}
+
+-- Are these supposed to be protected too?
+
+def add : complex → complex → complex :=
+λ z w, { re :=z.re+w.re, im:=z.im+w.im}
+
+def neg : complex → complex :=
+λ z, { re := -z.re, im := -z.im}
+
+def mul : complex → complex → complex :=
+λ z w, { re := z.re*w.re - z.im*w.im,
+         im := z.re*w.im + z.im*w.re}
+
+def norm_squared : complex → real :=
+λ z, z.re*z.re+z.im*z.im
+
+def inv : complex → complex :=
+λ z,  { re := z.re / norm_squared z,
+        im := -z.im / norm_squared z }
+
+instance : has_add complex := ⟨complex.add⟩
+instance : has_neg complex := ⟨complex.neg⟩
+instance : has_sub complex := ⟨λx y, x + - y⟩
+instance : has_mul complex := ⟨complex.mul⟩
+instance : has_inv complex := ⟨complex.inv⟩
+instance : has_div complex := ⟨λx y, x * y⁻¹⟩
+
+-- I was initially astounded to find that at some point there was a typo in has_div but
+-- this didn't cause any problems at all. I have since understood what is
+-- going on: "/" is never used in the field axioms, only ^{-1} .
+
+-- These are very useful for proofs in the library so I make them local simp lemmas.
```

Comment:
No harm in having these as global simp lemmas.

Reply:
I am slightly reluctant to overload simp, especially as I don't really understand how it works.